### PR TITLE
Backup

### PR DESCRIPTION
**Pull request by dziq45**

### CODE REVIEW

File: client/src/app/components/page/task-detail/task-detail.component.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    this.id = this.route.snapshot.paramMap.get('id');
+    if(this.id === 'newTask'){
+      this.updateMode=false;
+      this.buttonText='Add task';
+    }
+    this.taskForm = this.fb.group({
+      title:[''],
+      description:['']
+    });
+    this.ts.getTaskById(this.id).subscribe(task=>{
```
Comment (time2funk) [inline client/src/app/components/page/task-detail/task-detail.component.ts:41]: I think you need to search task data only if task id exist!
if not, than we creating a new one

File: client/src/app/components/page/task-detail/task-detail.component.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    })
+  }
+  onSubmit(){
+    if(this.updateMode){
+      const updatedTask:TaskUpdate = {
+        _id:this.id,
+        title: this.taskForm.value.title,
+        description:this.taskForm.value.description
+      }
+      this.ts.updateTask(updatedTask).subscribe();
```
Comment (time2funk) [inline client/src/app/components/page/task-detail/task-detail.component.ts:54]: you need to save all subscriptions and unsubscribe them onDestroy 
or you will catch memory leak

Comment (time2funk) [inline client/src/app/components/page/task-detail/task-detail.component.ts:54]: after update you can navigate to tasks page

File: client/src/app/components/page/task-detail/task-detail.component.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    else{
+      const newTask:TaskAdd={
+        title: this.taskForm.value.title,
+        description:this.taskForm.value.description,
+        createdBy: this.as.currentUserValue.userName,
+        parentId:null
+      }
+      this.ts.addTask(newTask).subscribe(newTask=>{
+        this.router.navigate(['/tasks/' + newTask._id]);
```
Comment (time2funk) [inline client/src/app/components/page/task-detail/task-detail.component.ts:64]: after we created a task need to save id
and switch to edit card

File: client/src/app/components/yes-no-modal/yes-no-modal.component.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { Component, OnInit, Input, EventEmitter } from '@angular/core';
+import {NgbActiveModal} from '@ng-bootstrap/ng-bootstrap';
+@Component({
+  selector: 'app-yes-no-modal',
+  templateUrl: './yes-no-modal.component.html',
+  styleUrls: ['./yes-no-modal.component.scss']
+})
+export class YesNoModalComponent implements OnInit {
+  answer:boolean = false;
```
Comment (time2funk) [inline client/src/app/components/yes-no-modal/yes-no-modal.component.ts:9]: public/private operators...

File: client/src/app/models/task.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+export interface Task {
+    _id: string;
+    title: string;
+    description: string;
+    createdAt: Date;
+    updatedAt: Date;
+    createdBy: string;
```
Comment (time2funk) [inline client/src/app/models/task.ts:7]: you can omit createdBy

File: client/src/app/models/task.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+export interface TaskUpdate{
+    _id:string;
+    title: string;
+    description:string;
+}
+export interface TaskAdd{
+    title: string;
+    description:string;
+    createdBy:string;
+    parentId:string;
```
Comment (time2funk) [inline client/src/app/models/task.ts:20]: parentId can be empty
**parentId:string** -> **parentId?:string**

File: client/src/app/services/task.service.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+  constructor(private apiService:ApiService) { }
+
+  getAllTasks():Observable<Task[]>{
+    return this.apiService.getAllTasks().pipe(
+      map((res:any)=>{
+        return res.result.data.map(item=>{
+          return <Task>item;
+          
+        })
```
Comment (time2funk) [inline client/src/app/services/task.service.ts:19]: **res.result.data.map(item=>{
    return <Task>item;
})**
->
**res.result.data.map(item => <Taks>item)**

File: client/src/app/AuthGuard.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import { Router, CanActivate, ActivatedRouteSnapshot, RouterStateSnapshot } from '@angular/router';
+import {AuthenticationService} from './services/authentication.service'
+@Injectable()
+export class AuthGuard implements CanActivate {
+
+    constructor(private router: Router, private as:AuthenticationService) {
+        
+     }
+
+    canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
```
[the remark this answers is not in the document]
Comment (dziq45) [reply client/src/app/AuthGuard.ts]: Done

File: client/src/app/components/page/list/list.component.ts [hunk 9]
```diff
@@ -1,15 +1,21 @@
...
   templateUrl: './list.component.html',
   styleUrls: ['./list.component.scss']
 })
 export class ListComponent implements OnInit {
-
-  constructor() { }
+  tasks:Task[];
+  constructor(private taskService: TaskService) { }
 
   ngOnInit() {
```
[the remark this answers is not in the document]
Comment (dziq45) [reply client/src/app/components/page/list/list.component.ts:13]: Done

File: client/src/app/components/page/login-page/login-page.component.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import {FormBuilder, FormGroup, FormControl, Validators} from '@angular/forms';
+import {AuthenticationService} from '../../../services/authentication.service';
+import { Router, ActivatedRoute } from '@angular/router';
+@Component({
+  selector: 'app-login-page',
+  templateUrl: './login-page.component.html',
+  styleUrls: ['./login-page.component.scss']
+})
+export class LoginPageComponent implements OnInit {
+  loginForm : FormGroup;
```
[the remark this answers is not in the document]
Comment (dziq45) [reply client/src/app/components/page/login-page/login-page.component.ts]: Done

File: client/src/app/components/page/navbar/navbar.component.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  templateUrl: './navbar.component.html',
+  styleUrls: ['./navbar.component.scss']
+})
+export class NavbarComponent implements OnInit {
+
+  currentUser: User;
+
+  constructor(public as: AuthenticationService) { }
+
+  ngOnInit() {
```
[the remark this answers is not in the document]
Comment (dziq45) [reply client/src/app/components/page/navbar/navbar.component.ts:15]: Done

File: client/src/app/components/page/task-detail/task-detail.component.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+import {ModalService} from '../../../services/modal.service';
+import {YesNoModalParams} from '../../../models/modals'
+import {TaskUpdate, TaskAdd} from '../../../models/task'
+@Component({
+  selector: 'app-task-detail',
+  templateUrl: './task-detail.component.html',
+  styleUrls: ['./task-detail.component.scss']
+})
+export class TaskDetailComponent implements OnInit {
+  updateMode:boolean = true;
```
[the remark this answers is not in the document]
Comment (dziq45) [reply client/src/app/components/page/task-detail/task-detail.component.ts]: Done

File: client/src/app/components/yes-no-modal/yes-no-modal.component.ts (hunk 4 above)
Comment (dziq45) [reply client/src/app/components/yes-no-modal/yes-no-modal.component.ts]: Added

File: client/src/app/models/task.ts (hunk 5 above)
Comment (dziq45) [reply client/src/app/models/task.ts]: Deleted createdBy

File: client/src/app/models/task.ts (hunk 6 above)
Comment (dziq45) [reply client/src/app/models/task.ts]: Now parentId can be null also

File: client/src/app/services/task.service.ts (hunk 7 above)
Comment (dziq45) [reply client/src/app/services/task.service.ts]: Refactored

File: client/src/app/components/page/task-detail/task-detail.component.ts (hunk 1 above)
Comment (dziq45) [reply client/src/app/components/page/task-detail/task-detail.component.ts]: You're right. Changed it

File: client/src/app/components/page/task-detail/task-detail.component.ts (hunk 2 above)
Comment (dziq45) [reply client/src/app/components/page/task-detail/task-detail.component.ts]: I read that a good practice is to use something like 
ngUnsubscribe = new Subject<void>();
subject.pipe(takeUntill(ngUnsubscribe)).subscribe(...)

so I did that to every subscription. I'm not sure if that's okay.

Comment (dziq45) [reply client/src/app/components/page/task-detail/task-detail.component.ts]: Changed

File: client/src/app/components/page/task-detail/task-detail.component.ts (hunk 3 above)
Comment (dziq45) [reply client/src/app/components/page/task-detail/task-detail.component.ts]: So I deleted this navigation to created task page and now we just swiched to editMode saving task's id.

File: client/src/app/models/task.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+export interface TaskUpdate{
+    _id:string;
+    title: string;
+    description:string;
+}
+export interface TaskAdd{
+    title: string;
+    description:string;
+    createdBy:string;
+    parentId:string|null;
```
Comment (time2funk) [inline client/src/app/models/task.ts:19]: you could use **?** operator to make this parameter optional
**parentId?:string**
this means that we can omit parentId parameter

File: client/src/app/components/page/list/list.component.ts [hunk 14]
```diff
@@ -1,15 +1,24 @@
...
   styleUrls: ['./list.component.scss']
 })
 export class ListComponent implements OnInit {
-
-  constructor() { }
+  public tasks:Task[];
+  private ngUnsubscribe = new Subject<void>();
+  constructor(private taskService: TaskService) { }
 
   ngOnInit() {
```
Comment (time2funk) [inline client/src/app/components/page/list/list.component.ts:16]: ngOnInit also can be public

File: client/src/app/components/page/list/list.component.ts [hunk 15]
```diff
@@ -1,15 +1,24 @@
...
+    this.getAllTasks();
+  }
+  public getAllTasks(){
+    this.taskService.getAllTasks().pipe(takeUntil(this.ngUnsubscribe)).subscribe(tasks=>{
```
Comment (time2funk) [inline client/src/app/components/page/list/list.component.ts:20]: please implement OnDestroy method in all paces where you have subscriptions
and complete your this.ngUnsubscribe there

File: client/src/app/models/task.ts (hunk 13 above)
Comment (dziq45) [reply client/src/app/models/task.ts]: Changed

File: client/src/app/components/page/list/list.component.ts (hunk 14 above)
Comment (dziq45) [reply client/src/app/components/page/list/list.component.ts]: Done

File: client/src/app/components/page/list/list.component.ts (hunk 15 above)
Comment (dziq45) [reply client/src/app/components/page/list/list.component.ts:20]: I added
```
    ngOnDestroy() {
        this.ngUnsubscribe.next()
        this.ngUnsubscribe.complete()
    }
```

File: client/src/app/components/page/navbar/navbar.component.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  public ngOnInit() {
+    this.as.currentUser.pipe(takeUntil(this.ngUnsubscribe)).subscribe(x => this.currentUser = x);
+  }
+  public onLogClick(){
+  
+  }
+  public onLogOut(){
+    this.as.logout();
+  }
+  public onDestroy(){
```
Comment (time2funk) [inline client/src/app/components/page/navbar/navbar.component.ts:27]: ngOnDestroy *

File: client/src/app/components/page/task-detail/task-detail.component.ts [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    modalRef.result.then((result) => {
+      if (result) {
+        this.ts.deleteTask(this.id).pipe(takeUntil(this.ngUnsubscribe)).subscribe(res => {
+          console.log(res)
+          this.router.navigate(['/tasks']);
+        })
+      }
+    });
+  }
+  public onDestroy(){
```
Comment (time2funk) [inline client/src/app/components/page/task-detail/task-detail.component.ts:94]: ngOnDestroy *

File: client/src/app/components/page/list/list.component.ts (hunk 15 above)
Comment (time2funk) [reply client/src/app/components/page/list/list.component.ts:20]: No, you added onDestroy, but need to be ngOnDestroy *

File: client/src/app/components/page/list/list.component.ts [hunk 18]
```diff
@@ -1,15 +1,28 @@
...
+import {TaskService} from '../../../services/task.service';
+import {Task} from '../../../models/task';
+import { Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
 @Component({
   selector: 'app-list',
   templateUrl: './list.component.html',
   styleUrls: ['./list.component.scss']
 })
 export class ListComponent implements OnInit {
```
Comment (time2funk) [inline client/src/app/components/page/list/list.component.ts:11]: implements OnInit, onDestroy {

File: client/src/app/components/page/navbar/navbar.component.ts [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+import {AuthenticationService} from '../../../services/authentication.service';
+import {User} from '../../../models/user';
+import { Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+@Component({
+  selector: 'app-navbar',
+  templateUrl: './navbar.component.html',
+  styleUrls: ['./navbar.component.scss']
+})
+export class NavbarComponent implements OnInit {
```
Comment (time2funk) [inline client/src/app/components/page/navbar/navbar.component.ts:11]: implements OnInit, onDestroy {

File: client/src/app/components/page/task-detail/task-detail.component.ts [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+import { YesNoModalParams } from '../../../models/modals'
+import { TaskUpdate, TaskAdd } from '../../../models/task'
+import { Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+@Component({
+  selector: 'app-task-detail',
+  templateUrl: './task-detail.component.html',
+  styleUrls: ['./task-detail.component.scss']
+})
+export class TaskDetailComponent implements OnInit {
```
Comment (time2funk) [inline client/src/app/components/page/task-detail/task-detail.component.ts:17]: implements OnInit, onDestroy {

File: client/src/app/components/page/list/list.component.ts (hunk 15 above)
Comment (dziq45) [reply client/src/app/components/page/list/list.component.ts:20]: My bad, changed it.

File: client/src/app/components/page/navbar/navbar.component.ts (hunk 16 above)
Comment (dziq45) [reply client/src/app/components/page/navbar/navbar.component.ts]: Changed

File: client/src/app/components/page/list/list.component.ts (hunk 18 above)
Comment (dziq45) [reply client/src/app/components/page/list/list.component.ts]: Fixed

File: client/src/app/components/page/navbar/navbar.component.ts (hunk 19 above)
Comment (dziq45) [reply client/src/app/components/page/navbar/navbar.component.ts]: Fixed

File: client/src/app/components/page/task-detail/task-detail.component.ts (hunk 20 above)
Comment (dziq45) [reply client/src/app/components/page/task-detail/task-detail.component.ts]: Fixed

**Pull request merged**